packages/context/test/acceptance/extension-point.ts

```diff
@@ -0,0 +1,215 @@
+    return await binding.getValue(extensionContext);
+  }
+}
+
+describe('Extension point', () => {
+  let ctx: Context;
+  beforeEach('given a context', createContext);
+
+  interface AuthenticationStrategy {
+    authenticate(credentials: any): Promise<boolean>;
+    config: any;
+  }
+
+  class LocalStrategy implements AuthenticationStrategy {
+    constructor(@inject('config') public config: any) {}
+
+    authenticate(credentials: any) {
+      return Promise.resolve(true);
+    }
+  }
+
+  class LDAPStrategy implements AuthenticationStrategy {
+    constructor(@inject('config') public config: any) {}
+
+    authenticate(credentials: any) {
+      return Promise.resolve(true);
+    }
+  }
+
+  class OAuth2Strategy implements AuthenticationStrategy {
+    constructor(@inject('config') public config: any) {}
+
+    authenticate(credentials: any) {
+      return Promise.resolve(true);
+    }
+  }
+
+  class AuthenticationManager extends ExtensionPoint<AuthenticationStrategy> {
+    constructor(
+      @inject('$context') context: Context,
+      @inject('authentication.config') config: ExtensionPointConfig,
+    ) {
+      super('authentication.strategies', context, config);
+    }
+
+    async authenticate(strategy: string, credentials: any): Promise<boolean> {
+      const ext: AuthenticationStrategy = await this.getExtension(strategy);
+      return ext.authenticate(credentials);
+    }
+  }
+
+  it('lists all extensions', async () => {
+    const authManager = await ctx.get('authentication.manager');
+    const extBindings = authManager.getExtensionBindings();
+    expect(extBindings.length).to.eql(3);
+  });
+
+  it('gets an extension by name', async () => {
+    const authManager: AuthenticationManager = await ctx.get(
+      'authentication.manager',
+    );
+    const binding = authManager.getExtensionBinding('ldap');
+    expect(binding.key).to.eql('authentication.strategies.ldap');
+    expect(binding.valueConstructor).to.exactly(LDAPStrategy);
+  });
+
+  it('gets an extension instance by name', async () => {
+    const authManager: AuthenticationManager = await ctx.get(
+      'authentication.manager',
+    );
+    const ext = await authManager.getExtension('ldap');
+    expect(ext.config).to.eql({
+      url: 'ldap://localhost:1389',
+    });
+    const result = await ext.authenticate({});
+    expect(result).to.be.true();
+  });
+
+  it('delegates to an extension', async () => {
+    const authManager: AuthenticationManager = await ctx.get(
+      'authentication.manager',
+    );
+    const result = await authManager.authenticate('local', {
+      username: 'my-user',
+      password: 'my-pass',
+    });
+    expect(result).to.be.true();
+  });
+
+  function createContext() {
+    ctx = new Context();
+
+    ctx.bind('$context').to(ctx);
+    // Register the extension point
+    ctx.bind('authentication.manager').toClass(AuthenticationManager);
+    ctx.bind('authentication.config').to({
+      extensions: {
```

Comment:
Can we avoid this extra level of nesting (config -> extensions -> local) and go for shorter config -> local?